scipy/optimize/_slsqp_py.py

```diff
@@ -439,26 +457,53 @@ def cjac(x, *args):
 
             # Print the status of the current iterate if iprint > 2
             if iprint >= 2:
-                print("%5i %5i % 16.6E % 16.6E" % (majiter, sf.nfev,
-                                                   fx, linalg.norm(g)))
+                print(
+                    f"{majiter:5} {sf.nfev:5} {fx:16.6E} "
+                    f"{linalg.norm(g):16.6E}"
+                )
 
         # If exit mode is not -1 or 1, slsqp has completed
         if abs(mode) != 1:
             break
 
         majiter_prev = int(majiter)
 
+    # obtain kkt multipliers from internals of SLSQP.
+    # Alternatively SLSQP needs to be called again to update the internals
+    # of the working array, w, whose first m entries are the multipliers.
+    im = 1
+    il = im + la
+    ix = il + (n1*n)//2 + 1
+    ir = ix + n - 1
+    _kkt_mult = np.abs(w[ir: ir + m])
```

Comment:
Hi. I've been testing this code out in the context of an optimal control problem where I need to extract the KKT multipliers. The code has been very helpful, but taking the absolute value here gives the wrong results. Is there a reason this is done?

Reply:
Do you have (minimal) test cases for which you know the code gives incorrect output? It'd be great to include those to see where remaining gaps in the implementation lie.

Reply:
Unfortunately nothing small and easy to check, the optimal control problem requires a whole mess of other code to set up.

I guess I would first ask why the absolute value is needed? If it is just to match the previous implementation where SQP is called twice then it is likely that the previous implementation wasn't quite correct. If I remember correctly, the KKT multipliers for _equality_ constraints can be positive or negative, but those for _inequality_ constraints have to be positive (or negative, depending on how you set the problem up).

Reply:
One way to see that the taking `abs` can't be right is to interpret the multipliers as the sensitivity of the objective function w.r.t. changes in the right hand side of the constraints. 

That is, suppose the equality constraint function is $g(x) = 0$. Suppose we find the optimal value of the objective function $f^\ast $, and the KKT multiplier corresponding with this constraint is $\lambda$. If we were to perturb the constraint to $g(x) = \epsilon$, where $\epsilon$ is small, then we would expect the new optimal value of the objective function to be $f_h^\ast \approx f^\ast + \epsilon \lambda $.

The problem is that we can write the same constraints as either $g(x) = 0$ or $-g(x) = 0$. When we perturb the two RHSs by $+\epsilon$, the change in the objective function will be in different directions - that is, the signs of $\lambda$ should be opposite. I'll illustrate this in the example below.

The reason why the Lagrange multipliers corresponding with the inequality constraints must be positive is that if we are solving a minimization problem with an inequality constraint $h(x) \geq 0$, increasing the right hand side by $\epsilon$ can _only_ make the objective function value worse, because we are tightening the constraint (making it more restrictive). For a minimizaiton problem, making it worse means making it more positive, so if it was $f^\ast$ before the perturbation, it becomes $f_h^\ast \approx f^\ast + \epsilon \mu$ with positive KKT multiplier $\mu$. (We can't multiplying both sides of $h(x) \geq 0$ by $-1$ as we did with the equality constraint because that would also change the direction of the inequality sign, and our problem is restricted to constraints expressed using a $\geq$ sign.)

Reply:
@jaharris85 thanks for your help finding the indices of the KKT multipliers. Could you tell us why you recommended taking the absolute value after getting the values from `w`? They seem to be correct without the absolute value.

Reply:
The intention was to line it up with the original multiplier values returned by the the Fortran code. Line 415 contains the multiplier update:
```
h3 = ABS(r(j))
mu(j) = MAX(h3,(mu(j)+h3)/two) 
```
where r is the more accurate version of the multipliers. So mu(j) must be positive since it is the max of something positive. I don't understand the intention of the second line though since that is what introduces the error in the multipliers.

Has anyone confirmed whether the original multipliers were always positive?

Reply:
From the perspective of the algorithm, the multipliers associated with inequality constraints  must be positive to satisfy the first order necessary conditions. See 2.3.1 of [A Software Package for Sequential Quadratic Programming] (https://degenerateconic.com/uploads/2018/03/DFVLR_FB_88_28.pdf). (But were you asking about whether that is actually the case in the code? All I know is it _should_ be.)